src/GuiBase/Timeline/Timeline.cpp

```diff
@@ -0,0 +1,505 @@
+        }
+    }
+}
+
+void Timeline::onChangingKeyFrame( size_t i ) {
+    if ( m_current.m_value )
+    {
+        Scalar time = m_current.m_value->getTimes()[i];
+        m_current.insertKeyFrame( time );
+        emit keyFrameChanged( i );
+    }
+}
+
+void Timeline::onMovingKeyFrame( size_t i, Scalar time1 ) {
+    if ( m_current.m_value )
+    {
+        if ( !Ra::Core::Math::areApproxEqual( m_current.m_value->getTimes()[i], time1 ) )
+        {
+            m_current.m_value->moveKeyFrame( i, time1 );
+            m_current.updateKeyFrame( time1 );
+            emit keyFrameMoved( i, time1 );
+        }
+    }
+}
+
+void Timeline::onMovingKeyFrames( size_t first, Scalar offset ) {
+    if ( m_current.m_value )
+    {
+        auto times        = m_current.m_value->getTimes();
+        const Scalar time = times[first];
+        if ( offset < 0 )
+        {
+            for ( int i = 0; i < times.size(); ++i )
+            {
+                const Scalar t = times[i];
+                if ( t >= time ) { m_current.m_value->moveKeyFrame( i, t + offset ); }
+            }
+        }
+        else // go from the end to ensure we do not mess up keyframes
+        {
+            for ( int i = times.size() - 1; i >= 0; --i )
+            {
+                const Scalar t = times[i];
+                if ( t >= time ) { m_current.m_value->moveKeyFrame( i, t + offset ); }
+            }
+        }
+        m_current.updateKeyFrame( time );
+        emit keyFramesMoved( first, offset );
+    }
+}
+
+void Timeline::on_comboBox_attribute_currentIndexChanged( const QString& arg1 ) {
+    onClearKeyFrames();
+    std::vector<Ra::Core::Animation::KeyFramedValueController> list;
+    const QStringList names = arg1.split( "::" );
+
+    // prints a warning message for the attribute and returns
+#define LOG_AND_RETURN                                                          \
+    LOG( logWARNING ) << "[Timeline] Error: attribute \"" << arg1.toStdString() \
+                      << "\"'s name is not conform.";                           \
+    return
+
+    // Checks if the given vector of KeyFramedValues contains a KeyFramedValue
+    // with the given name.
+    // If so, set the current frame to this KeyFramedValue;
+    // If not, prints a warning message for the attribute and returns.
+#define GET_KEYFRAMEDVALUE( list, name )                                                 \
+    {                                                                                    \
+        auto it = std::find_if( list.begin(), list.end(), [&name]( const auto& frame ) { \
+            return frame.m_name == name;                                                 \
+        } );                                                                             \
+        if ( it == list.end() ) { LOG_AND_RETURN; }                                      \
+        m_current = *it;                                                                 \
```

Comment:
If I understand well the code, here, the controller (`m_current`) is always initialized by a member wise affectation. What ensures that the pointer to a `KeyFrame` copied this way has longer lifetime that its alias into `m_current` ?

Reply:
nothing can ensure that, that's true.
Nevertheless, such keyframes must be registered into the Timeline, which checks for Entity, Component and RO removal, which somehow limits dead-ends.
The thing is I didn't find another way than raw pointer.

Reply:
shared_ptr ?

Reply:
nope, this won't work since it will be created by giving the address of the KeyFrame, which might not be stored in a shared_ptr on the data side, and thus when destroying the shared_ptr it will also destroy the KeyFrame while it shouldn't

Reply:
Couldn’t you store a shared_ptr on the data side ?